File: web/api/schemas/portfolio.py
"""Portfolio, stock, and asset response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from web.api.schemas.market import KLineItem
from web.api.schemas.strategy import StrategySignal


class StockDetail(BaseModel):
    symbol: str
    name: str
    industry: str
    sector: str


class FinancialData(BaseModel):
    period: str
    roe: float | None = None
    gross_margin: float | None = None
    net_margin: float | None = None
    debt_equity: float | None = None
    net_profit: float | None = None
    revenue: float | None = None
    profit_growth: float | None = None


class DCFParams(BaseModel):
    fcf: float = Field(..., description="自由现金流(亿)")
    growth_rate: float = Field(default=0.05, description="增长率")
    terminal_growth: float = Field(default=0.03)
    discount_rate: float = Field(default=0.08)
    shares: float = Field(..., description="总股本(亿)")


class DCFResult(BaseModel):
    intrinsic_value: float
    current_price: float
    safety_margin: float
    verdict: str


class StockResponse(BaseModel):
    basic: StockDetail
    financials: list[FinancialData]
    buffett_result: dict[str, Any] | None = None
    signals: list[StrategySignal]
    kline: list[KLineItem]
    dcf: DCFResult | None = None


class StockListItem(BaseModel):
    symbol: str
    name: str = ""
    industry: str = ""
    sector: str = ""
    price: float | None = None
    change_pct: float | None = None
    pe_ttm: float | None = None
    pb: float | None = None
    total_mv: float | None = None
    buffett_score: float | None = None
    roe: float | None = None
    gross_margin: float | None = None
    signal_score: float | None = None
    signal: str = "hold"
    buy_signals: int = 0
    signal_count: int = 0
    top_strategy: str = ""
    updated_at: str = ""


class StockListResponse(BaseModel):
    stocks: list[StockListItem]
    total: int
    limit: int
    updated_at: str = ""


class PositionItem(BaseModel):
    code: str
    name: str = ""
    volume: int
    avg_cost: float
    current_price: float
    market_value: float
    pnl: float
    pnl_pct: float


class AccountInfo(BaseModel):
    total_asset: float
    cash: float
    frozen_cash: float
    market_value: float


class OrderRequest(BaseModel):
    code: str
    price: float
    volume: int
    side: str = Field(..., description="buy/sell")


class OrderItem(BaseModel):
    order_id: str
    code: str
    side: str
    price: float
    volume: int
    filled_volume: int
    status: str
    created_at: str


class AssetOverviewItem(BaseModel):
    asset_type: str
    label: str
    enabled: bool = True
    data_source: str = "unknown"
    data_source_detail: str = ""
    research_ready: bool = False
    tradable: bool = False
    universe_size: int = 0
    error: str = ""


class AssetOverviewResponse(BaseModel):
    items: list[AssetOverviewItem] = []
    total: int = 0

File: web/frontend/src/api/modules/portfolio.ts
import { get, post } from "../client";
import type { PortfolioBalance, PortfolioNavPoint, PortfolioPosition, PortfolioSummary, PortfolioTrade } from "../types";

export const portfolioApi = {
  portfolioPositionRows: () => get<{ positions: any[] }>("/api/portfolio/positions"),
  portfolioPositions: async () => {
    const data = await get<{ positions: any[] }>("/api/portfolio/positions");
    return (data.positions || []).map((p) => ({
      symbol: p.code,
      name: p.name,
      shares: p.volume,
      cost: p.avg_cost,
      price: p.current_price,
      pnl: p.pnl,
      pnl_pct: (p.pnl_pct || 0) / 100,
    })) as PortfolioPosition[];
  },
  portfolioBalance: async () => {
    const data = await get<any>("/api/portfolio/balance");
    const initial = 1_000_000;
    const total = Number(data.total_asset || 0);
    const pnl = total - initial;
    return {
      cash: Number(data.cash || 0),
      total_value: total,
      total_pnl: pnl,
      total_pnl_pct: initial > 0 ? pnl / initial : 0,
    } as PortfolioBalance;
  },
  portfolioOrder: (order: { symbol: string; side: "buy" | "sell"; shares: number; price?: number }) =>
    post<any>("/api/portfolio/order", {
      code: order.symbol,
      side: order.side,
      volume: order.shares,
      price: order.price ?? 0,
    }),
  portfolioNav: () => get<{ nav: PortfolioNavPoint[] }>("/api/portfolio/nav"),
  portfolioTrades: (limit = 50) => get<{ trades: PortfolioTrade[]; total: number }>(`/api/portfolio/trades?limit=${limit}`),
  portfolioSummary: () => get<PortfolioSummary>("/api/portfolio/summary"),
  portfolioRefresh: () => post<any>("/api/portfolio/refresh"),
};
